rfc/rfc-73/rfc-73.md

```diff
@@ -0,0 +1,416 @@
+of potential failures increase. A failure in one table can cascade and affect the entire transaction. In case of
+failures, rolling back changes becomes more intricate in multi-table scenarios. Ensuring data consistency across tables
+during rollbacks, especially when there are inter-table dependencies, introduces additional challenges. Finally, the
+current concurrency control implementation is not designed to handle multi-table transactions. Hence, this RFC proposes
+a new design to support multi-table transactions.
+
+## Design
+
+First of all, let us discuss the goals and non-goals which will help in understanding the design considerations better.
+
+### Goals
+
+1. **Atomicity Across Tables:** Ensure that a set of changes spanning multiple tables either fully completes or fully
+   rolls back. No partial commits should occur.
+2. **Consistency:** Ensure that, after every transaction, all involved tables are in a consistent state. Avoid dirty
+   reads, non-repeatable reads, handle read-write and write-write conflicts.
+3. **Isolation:** Multiple concurrent transactions should not interfere with each other. One transaction shouldn't see
+   the intermediate states of another ongoing transaction.
+4. **Durability:** Once a multi-table transaction is committed, the changes should be permanent and recoverable, even
+   after system failures.
+5. **Performance:** The multi-table transaction mechanism should introduce minimal overhead. Its performance impact on
+   typical Hudi operations should be kept low. By corollary, locking (in case of OCC) cannot be coarser than file level.
+6. **Monitoring:** Integration with tools such as hudi-cli to track the status and health of transactions. Separate
+   transaction metrics. Also, provide comprehensive logs for debugging transaction failures or issues.
+7. **Integration with Current Hudi Features:** The multi-table transaction should be seamlessly integrated with existing
+   Hudi features and should not break existing APIs or workflows. In
+   particular, `TransacationManager`, `LockManager`, `ConflictResolutionStrategy` APIs should work as they do today with
+   single table.
+8. **Configurability:** Allow users to configure the behavior of multi-table transactions, e.g., concurrency control
+   mechanism, timeout durations, etc.
+9. **Recovery**: Rollbacks, savepoint and restore should be supported.
+
+### Non-goals
+
+1. **Cross-Database Transactions:** Transactions spanning multiple databases might be too complex as an initial target.
+2. **Granular Record Locking:** Instead of locking at the granularity of individual records, coarse-grained locking (
+   like at the file level) might be more practical to start with. But, let's avoid table or partition level locking as
+   mentioned in the goals.
+3. **Distributed Transactions:** We are not going to consider transactions in the sense of distributed databases, and
```

Comment:
In other words, storage nodes hosting the table data, are expected to be durable for a single table write - thus separating the distributed consensus issues. is that tht what this means?

Reply:
Yes, that's a design assumption. Each storage node is trusted to durably and reliably handle writes for an individual table. There's an implicit expectation of fault-tolerance at the storage level. None of the storage nodes are holding replicas.
This is unlike distrbuted databases like Amazon Aurora or CorckroachDB that will use a distributed consensus algorithm or quorum membership to guarantee consistency.

Reply:
can we state the assumptions here clearly "Storage is assumed to be durable and resilient to distributed failures" or sth short like that?